File: shared/main/java/net/ccbluex/liquidbounce/file/configs/XRayConfig.java
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.file.configs;

import java.io.*;

import com.google.gson.JsonArray;
import com.google.gson.JsonElement;
import com.google.gson.JsonParser;

import net.ccbluex.liquidbounce.LiquidBounce;
import net.ccbluex.liquidbounce.api.minecraft.client.block.IBlock;
import net.ccbluex.liquidbounce.features.module.modules.render.XRay;
import net.ccbluex.liquidbounce.file.FileConfig;
import net.ccbluex.liquidbounce.file.FileManager;
import net.ccbluex.liquidbounce.utils.ClientUtils;

public class XRayConfig extends FileConfig
{

	/**
	 * Constructor of config
	 *
	 * @param file
	 *             of config
	 */
	public XRayConfig(final File file)
	{
		super(file);
	}

	/**
	 * Load config from file
	 *
	 * @throws IOException
	 */
	@Override
	protected void loadConfig() throws IOException
	{
		final XRay xRay = (XRay) LiquidBounce.moduleManager.get(XRay.class);

		if (xRay == null)
		{
			ClientUtils.getLogger().error("[FileManager] Failed to find xray module.");
			return;
		}

		final JsonArray jsonArray = new JsonParser().parse(new BufferedReader(new FileReader(getFile()))).getAsJsonArray();

		xRay.getXrayBlocks().clear();

		for (final JsonElement jsonElement : jsonArray)
			try {
				final IBlock block = LiquidBounce.wrapper.getFunctions().getBlockFromName(jsonElement.getAsString());

				if (xRay.getXrayBlocks().contains(block)) {
					ClientUtils.getLogger().error("[FileManager] Skipped xray block '{}' because the block is already added.", block.getRegistryName());
					continue;
				}

				xRay.getXrayBlocks().add(block);
			} catch (final Throwable throwable) {
				ClientUtils.getLogger().error("[FileManager] Failed to add block to xray.", throwable);
			}
	}

	/**
	 * Save config to file
	 *
	 * @throws IOException
	 */
	@Override
	protected void saveConfig() throws IOException
	{
		final XRay xRay = (XRay) LiquidBounce.moduleManager.get(XRay.class);

		if (xRay == null)
		{
			ClientUtils.getLogger().error("[FileManager] Failed to find xray module.");
			return;
		}

		final JsonArray jsonArray = new JsonArray();

		for (final IBlock block : xRay.getXrayBlocks())
			jsonArray.add(FileManager.PRETTY_GSON.toJsonTree(LiquidBounce.wrapper.getFunctions().getIdFromBlock(block)));

		final PrintWriter printWriter = new PrintWriter(new FileWriter(getFile()));
		printWriter.println(FileManager.PRETTY_GSON.toJson(jsonArray));
		printWriter.close();
	}
}

File: shared/main/java/net/ccbluex/liquidbounce/file/configs/XRayConfig.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.file.configs

import com.google.gson.JsonArray
import com.google.gson.JsonObject
import com.google.gson.JsonParser
import net.ccbluex.liquidbounce.LiquidBounce
import net.ccbluex.liquidbounce.LiquidBounce.wrapper
import net.ccbluex.liquidbounce.features.module.modules.render.XRay
import net.ccbluex.liquidbounce.file.FileConfig
import net.ccbluex.liquidbounce.file.FileManager
import net.ccbluex.liquidbounce.utils.ClientUtils
import net.ccbluex.liquidbounce.utils.misc.MiscUtils
import java.io.File
import java.io.IOException

/**
 * Constructor of config
 *
 * @param file
 * of config
 */
class XRayConfig(file: File) : FileConfig(file)
{
	/**
	 * Load config from file
	 *
	 * @throws IOException
	 */
	@Throws(IOException::class)
	override fun loadConfig()
	{
		val xRay = LiquidBounce.moduleManager[XRay::class.java] as XRay
		val json = JsonParser().parse(MiscUtils.createBufferedFileReader(file))

		xRay.xrayBlocks.clear()

		val func = wrapper.functions

		(try
		{
			val jsonObject = json.asJsonObject

			xRay.orbfuscatorBypass = jsonObject["orbfuscatorBypass"].asBoolean

			jsonObject["blocks"].asJsonArray
		}
		catch (e: IllegalStateException)
		{
			json.asJsonArray // Backward-compatibility
		}).mapNotNull { func.getBlockFromName(it.asString) }.forEach { block ->
			try
			{
				if (xRay.xrayBlocks.contains(block))
				{
					ClientUtils.logger.warn("[FileManager] Skipped xray block '{}' because the block is already added.", block.registryName)
					return@forEach
				}

				xRay.xrayBlocks.add(block)
			}
			catch (throwable: Throwable)
			{
				ClientUtils.logger.error("[FileManager] Failed to add block to xray.", throwable)
			}
		}
	}

	/**
	 * Save config to file
	 *
	 * @throws IOException
	 */
	@Throws(IOException::class)
	override fun saveConfig()
	{
		val xRay = LiquidBounce.moduleManager[XRay::class.java] as XRay

		val jsonObject = JsonObject()

		jsonObject.addProperty("orbfuscatorBypass", xRay.orbfuscatorBypass)

		val blocks = JsonArray()

		val func = wrapper.functions
		xRay.xrayBlocks.map(func::getIdFromBlock).forEach { blockID ->
			blocks.add(FileManager.PRETTY_GSON.toJsonTree(blockID))
		}

		jsonObject.add("blocks", blocks)

		val writer = MiscUtils.createBufferedFileWriter(file)
		writer.write(FileManager.PRETTY_GSON.toJson(jsonObject) + System.lineSeparator())
		writer.close()
	}
}
